File: immutabilite/Plan.java
import java.util.ArrayList;
import java.util.List;

public class Plan {
    public static List<String> replanifie(List<String> plan, String nouvelleVille, String avantVille) {
            int index = plan.indexOf(avantVille);
            List<String> planB = new ArrayList<>(plan);
            
            planB.add(index, nouvelleVille);
            return planB;
    }
}

File: immutabilite/Plan.scala
object Plan  {
    def replanifie(plan: List[String], nouvelleVille: String, avantVille: String): List[String] = {
        val index = plan.indexOf(avantVille)
        val villeAvant = plan.slice(0, index)
        val villeApres = plan.slice(index, plan.size)
        villeAvant.appended(nouvelleVille).appendedAll(villeApres)    
    } 
    
    
    
    def main(args: Array[String]): Unit = {
       
        val plan1 = List("Paris", "Nantes")
        println("Plan 1 " + plan1)

        println("Après replanifie")
        val plan2 = replanifie(plan1, "Rennes", "Nantes")
        println("Plan 2 " + plan2)

    }

    
}